library/swift/src/Envoy.swift

```diff
@@ -16,32 +16,53 @@ public final class Envoy: NSObject {
     return self.runner.isFinished
   }
 
+  /// Initialize a new Envoy instance using a typed configuration.
+  ///
+  /// - parameter config:   Configuration to use for starting Envoy.
+  /// - parameter logLevel: Log level to use for this instance.
+  /// - parameter engine:   The underlying engine to use for starting Envoy.
+  init(config: EnvoyConfiguration, logLevel: LogLevel = .info, engine: EnvoyEngine) {
+    self.engine = engine
+    self.runner = RunnerThread(config: .typed(config), logLevel: logLevel, engine: engine)
+    self.runner.start()
+  }
+
   /// Initialize a new Envoy instance using a string configuration.
   ///
-  /// - parameter configYAML: Configuration YAML to use for starting Envoy.
+  /// - parameter configYAML: Configuration yaml to use for starting Envoy.
   /// - parameter logLevel:   Log level to use for this instance.
   /// - parameter engine:     The underlying engine to use for starting Envoy.
   init(configYAML: String, logLevel: LogLevel = .info, engine: EnvoyEngine) {
     self.engine = engine
-    self.runner = RunnerThread(configYAML: configYAML, logLevel: logLevel, engine: engine)
+    self.runner = RunnerThread(config: .yaml(configYAML), logLevel: logLevel, engine: engine)
     self.runner.start()
   }
 
   // MARK: - Private
 
   private final class RunnerThread: Thread {
     private let engine: EnvoyEngine
-    private let configYAML: String
+    private let config: ConfigurationType
     private let logLevel: LogLevel
 
-    init(configYAML: String, logLevel: LogLevel, engine: EnvoyEngine) {
-      self.configYAML = configYAML
+    enum ConfigurationType {
```

Comment:
This is essentially a `oneof` that ends up being cleaner than having two optional properties